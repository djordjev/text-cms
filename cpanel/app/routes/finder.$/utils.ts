import { CreateParam } from '~/types';

const getCreateOpenModal = (params: URLSearchParams): CreateParam | null => {
  const param = params.get('action');

  if (param === 'new-file' || param === 'new-folder') return param;

  return null;
};

export { getCreateOpenModal };
