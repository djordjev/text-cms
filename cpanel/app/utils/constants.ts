export const COLORS = {
  BLACK: '#000',
  PRIMARY: '#003EB3'
};

export const BUTTON_ACTION = '_action';
