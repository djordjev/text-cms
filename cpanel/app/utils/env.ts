const readEnvVariable = <T = string>(name: string) => {
  const vars = process.env;

  const val = vars[name];
  if (!val) throw new Error(`Variable ${name} not supplied`);

  return val as T;
};

export { readEnvVariable };
