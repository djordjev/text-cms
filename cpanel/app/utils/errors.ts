const buildErrorResponse = (status = 404, message = 'Not found') => {
  return new Response(message, { status, statusText: message });
};

export { buildErrorResponse };
