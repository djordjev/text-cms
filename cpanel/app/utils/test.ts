import { Descendant } from 'slate';

export const TEST_EDITOR: Descendant[] = [
  {
    type: 'heading',
    children: [
      {
        text: 'HEADING'
      }
    ],
    level: 3,
    align: 'center'
  },
  {
    type: 'paragraph',
    children: [
      {
        text: 'A '
      },
      {
        text: 'line',
        italic: true
      },
      {
        text: ' of '
      },
      {
        text: 'text',
        color: '#e41111',
        bold: true
      },
      {
        text: ' in a '
      },
      {
        text: 'paragraph',
        bold: true
      },
      {
        text: '.'
      }
    ]
  }
];

export const buildRequest = (data: Record<string, string>, method = 'POST') => {
  const form = new URLSearchParams(data);

  return new Request('https://www.test.local', { body: form, method });
};
