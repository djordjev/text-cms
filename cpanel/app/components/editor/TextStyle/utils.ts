import { BaseEditor, Editor, Element } from 'slate';
import { ReactEditor } from 'slate-react';

import { CustomElement } from '~/components/editor/VariationEditor/types';

const getSelectedStyle = (editor: BaseEditor & ReactEditor) => {
  const { selection } = editor;

  if (!selection) return;

  const [match] = Array.from(
    Editor.nodes<CustomElement>(editor, {
      at: selection,
      match: (n) =>
        !Editor.isEditor(n) &&
        Element.isElement(n) &&
        (n.type === 'paragraph' || n.type === 'heading')
    })
  );

  const [nodeType] = match;

  return nodeType;
};

export { getSelectedStyle };
