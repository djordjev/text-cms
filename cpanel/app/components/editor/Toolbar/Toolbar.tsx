import React from 'react';
import { Editor, Element, Transforms } from 'slate';
import { useSlate } from 'slate-react';

import { TextActions } from '~/components/editor/TextActions';
import { TextAlignment } from '~/components/editor/TextAlignment';
import { TextStyle } from '~/components/editor/TextStyle';
import { ToggleStyle } from '~/components/editor/ToggleStyle';
import { DEFAULT_TEXT_COLOR } from '~/routes/editor.$fileId.($variationId)/constants';
import {
  Action,
  AnyAction,
  ClickActionPayload,
  CustomElement
} from '~/types/editor';

export interface ToolbarProps {
  onClick: (action: Action, value?: AnyAction) => void;
}

const Toolbar: React.FC<ToolbarProps> = (props) => {
  const { onClick } = props;

  // Hooks
  const editor = useSlate();

  // Setup
  const marks = Editor.marks(editor);
  const color = marks?.[Action.Color] ?? DEFAULT_TEXT_COLOR;

  // Styles
  const classes = `u-flex u-items-center u-bg-base-300 u-mb-2x u-p-2xs`;

  // Handlers
  const onToggleClick = (action: Action) => (value: boolean) => {
    onClick(action, value);
  };

  const onPickColor = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.currentTarget.value;
    onClick(Action.Color, value);
  };

  const onSetClickAction = (payload?: ClickActionPayload) => {
    onClick(Action.Click, payload);
  };

  const onStylePick = (toTransform: Partial<CustomElement>) => {
    Transforms.setNodes<Element>(editor, toTransform);
  };

  const onAlignmentChange = (toTransform: Partial<CustomElement>) => {
    Transforms.setNodes<Element>(editor, toTransform);
  };

  // Markup
  const renderDivider = () => (
    <div className="u-bg-neutral u-mx-2xs" style={{ height: 18, width: 1 }} />
  );

  return (
    <div className={classes}>
      <div className="u-flex u-items-center">
        <ToggleStyle
          action={Action.Bold}
          className="u-mr-1xs"
          onToggle={onToggleClick(Action.Bold)}
        />
        <ToggleStyle
          action={Action.Italic}
          className="u-mr-1xs"
          onToggle={onToggleClick(Action.Italic)}
        />
        <ToggleStyle
          action={Action.Underline}
          className="u-mr-1xs"
          onToggle={onToggleClick(Action.Underline)}
        />
        <ToggleStyle
          action={Action.Strikethrough}
          onToggle={onToggleClick(Action.Strikethrough)}
        />
      </div>

      {renderDivider()}

      <input
        className="u-bg-transparent u-mr-1x"
        onChange={onPickColor}
        type="color"
        value={color}
      />

      {renderDivider()}

      <TextStyle onChange={onStylePick} />

      {renderDivider()}

      <TextAlignment onChange={onAlignmentChange} />

      {renderDivider()}

      <TextActions onClick={onSetClickAction} />
    </div>
  );
};

export { Toolbar };
