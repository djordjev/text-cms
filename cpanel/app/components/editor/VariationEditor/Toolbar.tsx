import { Popover } from '@headlessui/react';
import { ColorResult } from '@uiw/color-convert';
import Block from '@uiw/react-color-block';
import classnames from 'classnames';
import React from 'react';
import { Editor, Element, Transforms } from 'slate';
import { useSlate } from 'slate-react';

import { TextStyle } from '~/components/editor/TextStyle';
import { ToggleStyle } from '~/components/editor/ToggleStyle';
import { DEFAULT_TEXT_COLOR } from '~/components/editor/VariationEditor/constants';

import { Action } from './types';

export interface ToolbarProps {
  onClick: (action: Action, value?: string | boolean) => void;
}

const Toolbar: React.FC<ToolbarProps> = (props) => {
  const { onClick } = props;

  // Hooks
  const editor = useSlate();

  // Setup
  const marks = Editor.marks(editor);
  const color = marks?.[Action.Color] ?? DEFAULT_TEXT_COLOR;

  // Styles
  const classesLayout = `u-bg-gray-1 u-border-b u-border-solid u-border-primary-100 u-mb-2x u-p-2xs`;
  const classes = classnames(classesLayout, 'u-flex u-items-center');

  const onToggleClick = (action: Action) => (value: boolean) => {
    onClick(action, value);
  };

  const onPickColor = (color: ColorResult) => {
    onClick(Action.Color, color.hex);
  };

  const onStylePick = (toTransform: any) => {
    Transforms.setNodes<Element>(editor, toTransform);
  };

  return (
    <div className={classes}>
      <ToggleStyle action={Action.Bold} onToggle={onToggleClick(Action.Bold)} />
      <ToggleStyle
        action={Action.Italic}
        onToggle={onToggleClick(Action.Italic)}
      />
      <ToggleStyle
        action={Action.Underline}
        onToggle={onToggleClick(Action.Underline)}
      />
      <ToggleStyle
        action={Action.Strikethrough}
        onToggle={onToggleClick(Action.Strikethrough)}
      />
      <Popover className="u-inline-block u-mx-1x">
        <Popover.Button>
          <div
            className="u-rounded"
            style={{ backgroundColor: color, height: 14, width: 14 }}
          />
        </Popover.Button>
        <Popover.Panel className="u-absolute u-z-10">
          <Block color={'#fff'} onChange={onPickColor} />
        </Popover.Panel>
      </Popover>

      <TextStyle onChange={onStylePick} value={'paragraph'} />
    </div>
  );
};

export { Toolbar };
