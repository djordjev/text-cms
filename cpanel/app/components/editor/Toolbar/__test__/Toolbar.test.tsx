import { createRemixStub } from '@remix-run/testing';
import { render } from '@testing-library/react';
import { createEditor } from 'slate';
import { Slate, withReact } from 'slate-react';

import { TEST_EDITOR } from '~/utils/test';

import { Toolbar, ToolbarProps } from '../Toolbar';

test('Toolbar', () => {
  const onClick = vi.fn();

  const props: ToolbarProps = { onClick };

  const editor = withReact(createEditor());

  const createComponent = () => {
    const RemixStub = createRemixStub([
      {
        path: '/',
        Component: () => (
          <Slate editor={editor} initialValue={TEST_EDITOR}>
            <Toolbar {...props} />
          </Slate>
        )
      }
    ]);

    return <RemixStub />;
  };

  const { container } = render(createComponent());

  expect(container.childNodes).toMatchSnapshot();
});
