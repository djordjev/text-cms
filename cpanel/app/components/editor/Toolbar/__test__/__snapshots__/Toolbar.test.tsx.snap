// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toolbar 1`] = `
NodeList [
  <div
    class="u-flex u-items-center u-bg-base-300 u-mb-2x u-p-2xs"
  >
    <div
      class="u-flex u-items-center"
    >
      <button
        class="u-p-1xs u-mr-1xs"
        type="button"
      >
        <svg
          class="tabler-icon tabler-icon-bold"
          fill="none"
          height="20"
          role="img"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7 5h6a3.5 3.5 0 0 1 0 7h-6z"
          />
          <path
            d="M13 12h1a3.5 3.5 0 0 1 0 7h-7v-7"
          />
        </svg>
      </button>
      <button
        class="u-p-1xs u-mr-1xs"
        type="button"
      >
        <svg
          class="tabler-icon tabler-icon-italic"
          fill="none"
          height="20"
          role="img"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11 5l6 0"
          />
          <path
            d="M7 19l6 0"
          />
          <path
            d="M14 5l-4 14"
          />
        </svg>
      </button>
      <button
        class="u-p-1xs u-mr-1xs"
        type="button"
      >
        <svg
          class="tabler-icon tabler-icon-underline"
          fill="none"
          height="20"
          role="img"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7 5v5a5 5 0 0 0 10 0v-5"
          />
          <path
            d="M5 19h14"
          />
        </svg>
      </button>
      <button
        class="u-p-1xs"
        type="button"
      >
        <svg
          class="tabler-icon tabler-icon-strikethrough"
          fill="none"
          height="20"
          role="img"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 12l14 0"
          />
          <path
            d="M16 6.5a4 2 0 0 0 -4 -1.5h-1a3.5 3.5 0 0 0 0 7h2a3.5 3.5 0 0 1 0 7h-1.5a4 2 0 0 1 -4 -1.5"
          />
        </svg>
      </button>
    </div>
    <div
      class="u-bg-neutral u-mx-2xs"
      style="height: 18px; width: 1px;"
    />
    <input
      class="u-bg-transparent u-mr-1x"
      type="color"
      value="#000000"
    />
    <div
      class="u-bg-neutral u-mx-2xs"
      style="height: 18px; width: 1px;"
    />
    <select
      class="u-select-sm u-select-bordered"
    >
      <option
        value="heading_1"
      >
        Heading 1
      </option>
      <option
        value="heading_2"
      >
        Heading 2
      </option>
      <option
        value="heading_3"
      >
        Heading 3
      </option>
      <option
        value="heading_4"
      >
        Heading 4
      </option>
      <option
        value="heading_5"
      >
        Heading 5
      </option>
      <option
        value="heading_7"
      >
        Heading 6
      </option>
      <option
        value="paragraph"
      >
        Paragraph
      </option>
    </select>
    <div
      class="u-bg-neutral u-mx-2xs"
      style="height: 18px; width: 1px;"
    />
    <div>
      <button
        aria-label="Left"
        class="u-p-1xs u-bg-accent u-rounded-md"
        type="button"
      >
        <svg
          class="tabler-icon tabler-icon-align-left"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 6l16 0"
          />
          <path
            d="M4 12l10 0"
          />
          <path
            d="M4 18l14 0"
          />
        </svg>
      </button>
      <button
        aria-label="Center"
        class="u-p-1xs"
        type="button"
      >
        <svg
          class="tabler-icon tabler-icon-align-center"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 6l16 0"
          />
          <path
            d="M8 12l8 0"
          />
          <path
            d="M6 18l12 0"
          />
        </svg>
      </button>
      <button
        aria-label="Right"
        class="u-p-1xs"
        type="button"
      >
        <svg
          class="tabler-icon tabler-icon-align-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 6l16 0"
          />
          <path
            d="M10 12l10 0"
          />
          <path
            d="M6 18l14 0"
          />
        </svg>
      </button>
    </div>
    <div
      class="u-bg-neutral u-mx-2xs"
      style="height: 18px; width: 1px;"
    />
    <button
      class="u-p-1xs"
      type="button"
    >
      <svg
        class="tabler-icon tabler-icon-link"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9 15l6 -6"
        />
        <path
          d="M11 6l.463 -.536a5 5 0 0 1 7.071 7.072l-.534 .464"
        />
        <path
          d="M13 18l-.397 .534a5.068 5.068 0 0 1 -7.127 0a4.972 4.972 0 0 1 0 -7.071l.524 -.463"
        />
      </svg>
    </button>
    <div
      style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
    />
  </div>,
]
`;
